Traffic editor (#20)
osrf/ros2multirobotbook: src/traffic-editor.md

# Traffic Editor

In this section, we shall describe the traffic-editor GUI and simulation tools.

## Introduction and Objectives

Traffic management of heterogeneous robot fleets is non-trivial. One of the
challenges with coordinated management arises from varying semantics in
information models used across fleets. Representations of waypoints, lanes,
charging/docking stations, restricted zones, infrastructure  systems such as
doors & lifts among others are subject to vendor's discretion. However,
standardized conventions that convey the capabilities and intentions of fleets
in a shared facility are quintessential for planning. Multi-agent participants
in other modes of transportation such as roadways collectively adhere to a set

> of rules and conventions which minimizes chaos. More importantly, they allow for

Reviewer: ```suggestion
of rules and conventions which minimize chaos. More importantly, they allow for
```

a new participant to readily integrate into the system by following the
prescribed rules. Existing agents can accommodate the new participant as its
behavior is apparent. Traffic conventions for multi-robot systems do not exist.
The objective of the `traffic_editor` is to fill this gap by expressing the
intentions of various fleets in a standardized, vendor neutral manner through a
graphical interface. Collated traffic information from different fleets can then

> be exported for planning and control. A secondary objective of the

Reviewer: ```suggestion
be exported for planning and control. A secondary objective and benefit of the
```

`traffic_editor` is to facilitate generation of 3D simulations worlds which
accurately reflect physical environments.

## Overview

> The `traffic_editor` [repository](https://github.com/osrf/traffic_editor) is home to the `traffic_editor` GUI and tools to auto-generate simulation worlds from GUI output. 

Reviewer: ```suggestion
The `traffic_editor` [repository](https://github.com/osrf/traffic_editor) is home to the `traffic_editor` GUI and tools to auto-generate simulation worlds from GUI output. The GUI is an easy to use interface which can be used to create and annotate 2D floor plans with robot traffic along with building infrastructure information. In the case there are existing floor plans of the environment, the `traffic_editor` is able to import those images for faster layout and floor generation.
```

> 

Reviewer: ```suggestion
```

The functionality of the GUI is to load and annotate 2D floor plans with robot
traffic along with building infrastructure information. While readily available
3D models of sites would ease the process running simulations, in practice these
are often unavailable or do not exist for distribution from facility owners. Floor
plans however are relatively easier to access. Hence, the `traffic_editor` is
designed to work with such 2D drawings to help build up accurate digital

> representations of facilities.

Reviewer: Suggest to delete this preceding paragraph. I moved the first sentence into the previous paragraph and also tried to re-frame the problem statement about 3D/2D models. Afaik, we can convert 3D BIM files if we need to as well but currently `traffic_editor` imports only image files (which could come from 2D or 3D CAD). Totally open to changes or alternate approaches on the paragraph(s)

Author: Deleted https://github.com/osrf/ros2multirobotbook/pull/20/commits/7246171e3476fc0bbbb5402dae5dd21e8d163165

tag with the appropriate `.building.yaml` file will set the editor to load and
apply any annotations to this file.

```yaml
building:
  filename: ""
name: ""
traffic_maps:
  {}
version: 1
```

Each `.building.yaml` file encompasses several attributes for each level in
the site as annotated by the user. A detailed description of the various
options are discussed in the subsequent section.
 
```yaml
levels:
  L1:
    doors:
      - []
    drawing:
      filename:
    fiducials:
    elevation: 0
    flattened_x_offset: 0
    flattened_y_offset: 0
    floors:
      - parameters: {}
        vertices: []
    lanes:
      - []
    layers:
      {}
    measurements:
      - []

    models:
      -{}
    vertices:
      {}
    walls:
      {}
lifts:
  {}
name: building

```

## GUI Layout
The layout of the `traffic_editor` comprises of a `Main Toolbar`, a `Working Area`
and a `Sidebar` as seen in the figure below.

![](images/traffic_editor/layout.png)

The editor operates in three modes as selectable from the `Edit mode` dropdown
list in the `Main Toolbar` or using shortkeys:
* **Building (Ctrl + B):** to annotate _walls_, floors (_polygons_), _doors_,

>   _measurements_ to set the scale of the drawing and _fiducials_ to align levels

Reviewer: ```suggestion
  _measurements_ to set the scale of the drawing, _fiducials_ to align levels
```

>   for multi-level simulations.  

Reviewer: ```suggestion
  for multi-level simulations and environment assets to add virtual models to your simulated environment.  
```

* **Traffic (Ctrl + T):** to annotate traffic lanes of various robots on each level
* **Scenario (Ctrl + E):** to define special regions of interest on a level. In
  larger facilities robot traffic may be limited to certain areas only. Hence,
  it may be more meaningful to generate simulation worlds of these special
  regions alone which can be specified in this mode.

Each mode has a unique set of tools while the following are common:
|                    Icon                    |  Name  | Shortkey |               Function               |
|:------------------------------------------:|:------:|:--------:|:------------------------------------:|
| ![](images/traffic_editor/icons/select.svg) | Select |   `Esc`  | Select an entity in the `Working Area` |
|  ![](images/traffic_editor/icons/move.svg)  |  Move  |    `m`   |  Move an entity in the `Working Area`  |
| ![](images/traffic_editor/icons/rotate.svg) | Rotate |    `r`   | Rotate an entity in the `Working Area` |

> The `SideBar` consists multiple tabs with various functionalities:

Reviewer: ```suggestion
The `SideBar` contains multiple tabs with various functionalities:
```

> * **levels:** to add a new level to the building using a floor plan

Reviewer: ```suggestion
* **levels:** to add a new level to the building. This can be done from scratch of by importing a floor plan image file
```

been added as reflected in the `levels` tab. A default scale fo `1px = 5cm` is
applied. The actual scale can be set by adding a measurement. Any offsets
applied to align levels will be reflected in the `X` and `Y` columns. Saving the
project will update the `tutorial.building.yaml` files as seen below.
```yaml

levels:
  L1:
    drawing:
      filename: office.png
    elevation: 0
    flattened_x_offset: 0
    flattened_y_offset: 0
    layers:
      {}
lifts:
  {}
name: building
```
### Adding a vertex
|                    Icon                    | Shortkey |
|:------------------------------------------:| :-------:| 
| ![](images/traffic_editor/icons/vertex.svg)| `v` |

A vertex is a fundamental component of multiple annotations. Walls,
measurements, doors, floor polygons and traffic lanes are created from two or
more vertices. To create a vertex, click on the vertex icon in the `Main
Toolbar` and anywhere on the canvas. The default attributes of a vertex are its
coordinates along with an empty name field. Additional attributes may be added
by clicking the `Add` button in the figure. Short descriptions of these are
presented below.

> * **is_partking_spot:** if true and if the waypoint is part of a traffic lane,

Reviewer: ```suggestion
* **is_parking_spot:** if true and if the waypoint is part of a traffic lane,
```

  the `rmf_fleet_adapter` will treat this as a _holding point_ during path

>   planning, ie, the robot is allowed to wait at this waypoint for an indefinite

Reviewer: ```suggestion
  planning, i.e., the robot is allowed to wait at this waypoint for an indefinite
```

  period of time.
* **is_charger:** if true and if the waypoint is part of a traffic lane, the

>   `rmf_fleet_adapter` will treat this as a charing station.

Reviewer: ```suggestion
  `rmf_fleet_adapter` will treat this as a charging station.
```

* **dock_name:** if specified and if the waypoint is part of a traffic lane, the
  `rmf_fleet_adapter` will issue an `rmf_fleet_msgs::ModeRequest` message with
  `MODE_DOCKING` and `task_id` equal to the specified name to the robot as it
  approaches this waypoint.

> * **workcell_name [deprecated]:** This attribute was previously used by the

Reviewer: If this is already deprecated and this is our first release of the documentation, should we even include this?

Author: That is a good point. I added this note since the gui still has this option. Perhaps we can remove it when a future release of traffic editor gets rid of this option?

Reviewer: sounds like a plan! :+1:

originally engineered to showcase shared lift access in a single floor
demonstration environment with a "mock" lift that receives lift commands and
transmits lift states but does not actually move between any different floors in
a building. However, as there may be interest such functionality for testing
single-floor hardware setups that seek to emulate multi-floor scenarios, these
properties were retained.

* **demo_mock_floor_name**: name of the floor that the robot is on while
  traversing the lane
* **demo_mock_lift_name**: name of the lift that is being entered or exited
  while the robot traverses the lane

To further explain these properties, consider this representation of a
navigation graph where numbers are waypoints and letters are lanes:
```
1 <---a---> 2 <---b---> 3

Waypoint 1 is on floor L1
Waypoint 2 is inside the "lift" named LIFT001
Waypoint 3 is on floor L3
The properties of edge "a" are:
    bidirectional: true
    demo_mock_floor_name: L1
    demo_mock_lift_name: LIFT001
The properties of edge "b" are:
    bidirectional: true
    demo_mock_floor_name: L3
    demo_mock_lift_name: LIFT001
```

> If the robot is to travel from 1 to 3, the `rmf_fleet_adapter` will request for

Reviewer: ```suggestion
If the robot is to travel from level 1 to 3, the `rmf_fleet_adapter` will request for
```

Author: Modified to "from waypoint 1 to waypoint 3" for better clarity. https://github.com/osrf/ros2multirobotbook/pull/20/commits/7246171e3476fc0bbbb5402dae5dd21e8d163165

> the "mock lift" to arrive at L1 when the robot approaches 1. With confirmation

Reviewer: ```suggestion
the "mock lift" to arrive at L1 when the robot approaches waypoint 1. With confirmation
```
